### PR TITLE
Use ShorthandSerializer for grid-template shorthand.

### PR DESCRIPTION
**Pull request by sammygill**
#### 4820e6a9d6603660f37396c83f290b9b44afd92b
<pre>
Use ShorthandSerializer for grid-template shorthand.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260397.">https://bugs.webkit.org/show_bug.cgi?id=260397.</a>
rdar://114097919.

Reviewed by Darin Adler.

The ShorthandSerializer already serializes the shorthand into its
shortest representation. When we get the property value for the computed
style, we can use the ShorthandSerializer instead.

In order to do this, the ShorthandSeralizer needs a new constructor that
can take a ComputedStyleExtractor. This will be used in commonSerializationChecks
to set the longhand values appropriately.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/grid-layout-properties-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::getPropertyValue const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::ShorthandSerializer):
(WebCore::ShorthandSerializer::commonSerializationChecks):
(WebCore::serializeShorthandValue):
* Source/WebCore/css/ShorthandSerializer.h:

Canonical link: <a href="https://commits.webkit.org/267280@main">https://commits.webkit.org/267280@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0c1f5d321a00a793df57ce8df07757cbe2d3e907

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17896 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15139 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16319 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19485 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16559 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17548 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16327 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16787 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13776 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18660 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14030 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14594 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21434 "11 flakes 123 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15019 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14759 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17992 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13016 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14580 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3863 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18949 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15175 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/CSSComputedStyleDeclaration.cpp [hunk 1]
```diff
@@ -105,6 +106,17 @@ const FixedVector<CSSPropertyID>& CSSComputedStyleDeclaration::exposedComputedCS
...
+    auto canUseShorthandSerializerForPropertyValue = [&]() {
+        switch (propertyID) {
+        case CSSPropertyGridTemplate:
+            return true;
+        default:
+            return false;
+        }
+    };
+    if (isShorthand(propertyID) && canUseShorthandSerializerForPropertyValue())
+        return serializeShorthandValue({ m_element, m_allowVisitedStyle, m_pseudoElementSpecifier }, propertyID);
```
Comment (nt1m) [inline Source/WebCore/css/CSSComputedStyleDeclaration.cpp]: I think `ComputedStyleExtractor::valueForPropertyInStyle` is a better place for this, we can also just re-use the already initialized `ComputedStyleExtractor` there too.

Comment (sammygill) [reply Source/WebCore/css/CSSComputedStyleDeclaration.cpp]: Looks like we would need to rework some things in order to do this since the serializer uses Strings and the extractor uses CSSValues

Comment (darinadler) [reply Source/WebCore/css/CSSComputedStyleDeclaration.cpp]: This is a big problem with our code for computed style — there’s no web platform reason that computed style property values need to be turned into CSS value objects, but there are internal WebKit reasons, like the way we use it in some HTML editing code.

But because the web platform doesn’t specify this, a lot of the CSS values that we create are arbitrary and all that matters from a web platform point of view is that they serialize to a string as specified. At some point we should fix that by eliminating the need to create CSS values from computed style. Then we can delete a ton of redundant computed style code and likely improve the correctness of all computed style shorthand serialization.

File: Source/WebCore/css/CSSComputedStyleDeclaration.cpp [hunk 2]
```diff
@@ -105,6 +106,17 @@ const FixedVector<CSSPropertyID>& CSSComputedStyleDeclaration::exposedComputedCS
 
 String CSSComputedStyleDeclaration::getPropertyValue(CSSPropertyID propertyID) const
 {
+    auto canUseShorthandSerializerForPropertyValue = [&]() {
+        switch (propertyID) {
+        case CSSPropertyGridTemplate:
```
Comment (darinadler) [inline Source/WebCore/css/CSSComputedStyleDeclaration.cpp:111]: Seems like we can do this for a **lot** more properties, maybe even most, in the longer term. Eventually I’d want this to be reversed in sense, with a list of the shorthand properties that **cannot** use the shorthand serializer. The downside is that `getPropertyCSSValue` for these properties will then be largely untested, since the web platform tests won’t exercise that code much any more.

File: Source/WebCore/css/ComputedStyleExtractor.h [hunk 3]
```diff
@@ -58,6 +58,8 @@ class ComputedStyleExtractor {
     ComputedStyleExtractor(Node*, bool allowVisitedStyle, PseudoId);
     ComputedStyleExtractor(Element*, bool allowVisitedStyle = false);
     ComputedStyleExtractor(Element*, bool allowVisitedStyle, PseudoId);
+    ComputedStyleExtractor(Element&, bool allowVisitedStyle = false);
+    ComputedStyleExtractor(Element&, bool allowVisitedStyle, PseudoId);
```
Comment (darinadler) [inline Source/WebCore/css/ComputedStyleExtractor.h]: Do we really need this? Is there an important reason to not just add a `&` at the call site?

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 4]
```diff
@@ -4452,6 +4462,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
     case CSSPropertyTextAnchor:
     case CSSPropertyVectorEffect:
         return svgPropertyValue(propertyID);
+
+    // grid-template is handled by ShorthandSeralizer
```
Comment (darinadler) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: To me this comment is confusing and doesn’t feel relevant.

The reason it’s OK to have the `grid-template` case return `nullptr` here is that somehow we know that it’s OK for `valueForPropertyInStyle` and hence for `getPropertyCSSValue` to return `nullptr` for this property. That’s what is non-obvious and thus that is what comment should be about, not about how `getPropertyValue` works.

What is the basis of that? How did we determine that we don’t have any code that tries to get a CSS value for this property?

File: Source/WebCore/css/ShorthandSerializer.h [hunk 5]
```diff
@@ -29,9 +29,11 @@
 namespace WebCore {
 
 class StyleProperties;
+class ComputedStyleExtractor;
```
Comment (darinadler) [inline Source/WebCore/css/ShorthandSerializer.h:31]: We sort these alphabetically.

File: Source/WebCore/css/ShorthandSerializer.cpp [hunk 6]
```diff
@@ -182,6 +185,20 @@ bool ShorthandSerializer::subsequentLonghandsHaveInitialValues(unsigned startInd
     return true;
 }
 
+bool ShorthandSerializer::commonSerializationChecks(const ComputedStyleExtractor& properties)
+{
+    ASSERT(length());
+
+    if (m_shorthand.id() == CSSPropertyAll)
+        return false;
```
Comment (darinadler) [inline Source/WebCore/css/ShorthandSerializer.cpp]: Why is this needed? What happens if you remove it? Can it be an assertion instead?

File: Source/WebCore/css/ShorthandSerializer.h [hunk 7]
```diff
@@ -29,9 +29,11 @@
...
 
 enum CSSPropertyID : uint16_t;
 
 String serializeShorthandValue(const StyleProperties&, CSSPropertyID);
+String serializeShorthandValue(const ComputedStyleExtractor&, CSSPropertyID);
```
Comment (darinadler) [inline Source/WebCore/css/ShorthandSerializer.h:36]: I’d suggest sorting these alphabetically too, because "why not".

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 4 above)
Comment (sammygill) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Yeah that's a good point and now is making me rethink whether removing this code is actually correct (doesn't seem like it anymore). Allowing `getPropertyValue` to use `ShorthandSerializer` is one thing but returning `nullptr` here has other implications and probably very unexpected for other callers
